src/control/ToolHandler.cpp

```diff
@@ -446,13 +483,20 @@ void ToolHandler::loadSettings() {
     }
 }
 
-void ToolHandler::pointCurrentToolToButtonTool() { this->currentTool = this->buttonSelectedTool; }
+void ToolHandler::pointCurrentToolToButtonTool(Button button) {
+    Tool* tool = getButtonTool(button);
+    if (this->activeTool == tool)
+        return;
+    this->activeTool = tool;
+
+    this->fireToolChanged();
+}
```

Comment:
@LittleHuba : This could be quite big performance improvement as this function is called in every call of `PenInputHandler::actionMotion` via `changeTool` which is basically called continuously while drawing for reasons I haven't figured out yet (see current discussion in https://github.com/xournalpp/xournalpp/pull/2393/files#diff-a03baeadae3022a320f793c7565ed67b183f6755149f324133430b629ee8956bL172). If you know why `changeTool` needs to be called throughout the Motion that would be great. If not I'll play around with it.

Reply:
Just gave my two cents on that comment. You are welcome to play around with that though. If you manage to reduce those calls (definitely possible) it would help a lot.

Reply:
I adapted the code according to the comments. It took a while but I think now `fireToolChanged()` is almost only called when one would expect.